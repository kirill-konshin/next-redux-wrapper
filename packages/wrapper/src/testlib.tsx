import {AppContext} from 'next/app';
import {applyMiddleware, createStore, AnyAction} from 'redux';
import promiseMiddleware from 'redux-promise-middleware';
import React, {Component} from 'react';
import renderer from 'react-test-renderer';

export const reducer = (state = {reduxStatus: 'init'}, action: AnyAction) => {
    switch (action.type) {
        case 'FOO': // sync
        case 'FOO_FULFILLED': // async
            return {reduxStatus: action.payload};
        default:
            return state;
    }
};

export const makeStore = (initialState: any) => createStore(reducer, initialState, applyMiddleware(promiseMiddleware));

class SyncPageBase extends Component<any> {
    public render() {
        const {store, ...props} = this.props;
        return (
            <div>
                {JSON.stringify(props)}
                {JSON.stringify(store.getState())}
            </div>
        );
    }
}

export class SyncPage extends SyncPageBase {
    public static getInitialProps({ctx}: AppContext) {
        ctx.store.dispatch({type: 'FOO', payload: 'foo'});
        return {custom: 'custom'};
    }
}

const someAsyncAction = {
    type: 'FOO',
    payload: new Promise(res => res('foo')),
};

export class AsyncPage extends SyncPageBase {
    public static async getInitialProps({ctx}: AppContext) {
        await ctx.store.dispatch(someAsyncAction);
        return {custom: 'custom'};
    }
}

export class NoStorePage extends SyncPageBase {
    public static async getInitialProps() {
        return {custom: 'custom'};
    }
}

/**
 * Creates a stub AppContext declared as any
 */
export function createAppContext(): any {
    return {
        ctx: {
            pathname: '/',
            req: {},
            res: {},
            query: {},
        },
    };
}

export async function verifyComponent(WrappedPage: any) {
    // this is called by Next.js
    const props = await WrappedPage.getInitialProps(createAppContext());

    expect(props.initialProps.custom).toBe('custom');
    expect(props.initialState.reduxStatus).toBe('foo');

    // this is called by Next.js
    const component = renderer.create(<WrappedPage {...props} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
}
