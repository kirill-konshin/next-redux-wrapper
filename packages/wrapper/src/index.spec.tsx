import * as React from 'react';
import renderer from 'react-test-renderer';
import withRedux from './index';
import {AsyncPage, makeStore, SyncPage, verifyComponent, createAppContext} from './testlib';

describe('store integration', () => {
    test('simple', async () => {
        const WrappedPage = withRedux(makeStore)(SyncPage);
        await verifyComponent(WrappedPage);
    });

    test('async', async () => {
        const WrappedPage = withRedux(makeStore)(AsyncPage);
        await verifyComponent(WrappedPage);
    });
});

describe('custom serialization', () => {
    test('custom state serialization on the server and deserialization on the client', async () => {
        class MyApp extends React.Component<any> {
            public render() {
                const {store} = this.props;
                return <div>{JSON.stringify(store.getState())}</div>;
            }
        }

        const WrappedPage = withRedux(makeStore, {
            serializeState: state => ({...state, serialized: true}),
            deserializeState: state => ({...state, deserialized: true}),
            debug: true,
        })(MyApp);

        const props = await WrappedPage.getInitialProps(createAppContext());
        expect(props.initialState.serialized).toBeTruthy();

        const component = renderer.create(<WrappedPage {...props} />);

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
