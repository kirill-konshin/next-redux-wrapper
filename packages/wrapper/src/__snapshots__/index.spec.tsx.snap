// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom serialization custom state serialization on the server and deserialization on the client 1`] = `
<div>
  {"deserialized":true,"serialized":true}
</div>
`;

exports[`store integration async 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"foo"}
</div>
`;

exports[`store integration simple 1`] = `
<div>
  {"isServer":true,"custom":"custom"}
  {"reduxStatus":"foo"}
</div>
`;
